File: hardening/week_grid_oracle.java
// Independent-runtime oracle: lays out a month with java.time.temporal.WeekFields under OpenJDK 21 and
// prints, for each day of that month, the week of the month and the day of the week that WeekFields
// assigns it. Arguments come in groups of four -- year, month, first weekday, minimal days in the first
// week -- so one process answers every month of one example. The only translation is of vocabulary: the
// first weekday arrives numbered as CPython's calendar module numbers it, Monday 0 through Sunday 6,
// and java.time.DayOfWeek numbers the same seven days Monday 1 through Sunday 7, so DayOfWeek.of takes
// the neighbouring value. Each group prints one line: the length of the month, then one tab-separated
// field per day holding the week of the month and the day of the week separated by a comma.
// Usage: java week_grid_oracle.java <year> <month> <first-weekday> <minimal-days> ...
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.time.temporal.WeekFields;

public class WeekGridOracle {
    public static void main(String[] args) {
        for (int at = 0; at < args.length; at += 4) {
            int year = Integer.parseInt(args[at]);
            int month = Integer.parseInt(args[at + 1]);
            WeekFields fields = WeekFields.of(DayOfWeek.of(Integer.parseInt(args[at + 2]) + 1),
                                              Integer.parseInt(args[at + 3]));
            int length = YearMonth.of(year, month).lengthOfMonth();
            StringBuilder line = new StringBuilder(Integer.toString(length));
            for (int day = 1; day <= length; day++) {
                LocalDate date = LocalDate.of(year, month, day);
                line.append('\t').append(date.get(fields.weekOfMonth()));
                line.append(',').append(date.get(fields.dayOfWeek()));
            }
            System.out.println(line);
        }
    }
}

File: hardening/week_grid_oracle.php
<?php
// Independent-runtime oracle: lays out a month with ICU's calendar through PHP's intl extension and
// prints, for each day of that month, the week of the month and the day of the week ICU assigns it.
// Arguments come in groups of four -- year, month, first weekday, minimal days in the first week -- so
// one process answers every month of one example. Two translations of vocabulary and nothing else: the
// month arrives numbered from one as CPython's calendar module numbers it and ICU numbers its months
// from zero, and the first weekday arrives numbered Monday 0 through Sunday 6 as that module numbers it
// while ICU numbers the same seven days Sunday 1 through Saturday 7. Each group prints one line: the
// length of the month, then one tab-separated field per day holding the week of the month and the day
// of the week separated by a comma.
// Usage: php week_grid_oracle.php <year> <month> <first-weekday> <minimal-days> ...
$arguments = array_slice($argv, 1);
for ($at = 0; $at < count($arguments); $at += 4) {
    $year = (int) $arguments[$at];
    $month = (int) $arguments[$at + 1] - 1;
    $calendar = IntlCalendar::createInstance(new DateTimeZone('UTC'), 'en_US');
    $calendar->setFirstDayOfWeek(((int) $arguments[$at + 2] + 1) % 7 + 1);
    $calendar->setMinimalDaysInFirstWeek((int) $arguments[$at + 3]);
    $calendar->setTime(0);
    $calendar->set($year, $month, 1);
    $length = $calendar->getActualMaximum(IntlCalendar::FIELD_DAY_OF_MONTH);
    $line = (string) $length;
    for ($day = 1; $day <= $length; $day++) {
        $calendar->set($year, $month, $day);
        $line .= "\t" . $calendar->get(IntlCalendar::FIELD_WEEK_OF_MONTH);
        $line .= "," . $calendar->get(IntlCalendar::FIELD_DAY_OF_WEEK);
    }
    echo $line . "\n";
}
